### ISSUE TITLE
WASM module downloads in real Chrome with http.FileServer but not Chromedp

### ISSUE THREAD

**Issue by Gregory-Ledray** [chromedp/chromedp#571]
Short on time? Skip here and read from there:
https://github.com/chromedp/chromedp/issues/571#issuecomment-582650524

Basically, for some reason my tests work in real Chrome but not when run in chromedp.

#### What versions are you running?

<pre>
$ go list -m github.com/chromedp/chromedp
github.com/chromedp/chromedp v0.3.1-0.20190619195644-fd957a4d2901
$ google-chrome --version
Google Chrome 79.0.3945.130
$ go version
go version go1.13.7 linux/amd64
</pre>

Full error which I'm getting in WSL with chromedp:
```
[wasmbrowsertest]: 2020/02/04 23:01:50 main.go:214: {"type":"object","subtype":"error","className":"TypeError","description":"Ty
peError: Could not download wasm module","objectId":"{\"injectedScriptId\":2,\"id\":1}","preview":{"type":"object","subtype":"er
ror","description":"TypeError: Could not download wasm module","overflow":false,"properties":[{"name":"stack","type":"string","v
alue":"TypeError: Could not download wasm module"},{"name":"message","type":"string","value":"Could not download wasm module"}]}
}
```
See the source code in chromium throwing this error: https://chromium.googlesource.com/chromium/src/+/70d7b68f1669306da4c713c7e4d4b0be97a13ab8/third_party/WebKit/Source/bindings/modules/v8/wasm/WasmResponseExtensions.cpp

And a tangentially related commit: https://chromium-review.googlesource.com/c/chromium/src/+/876103

## Reproduction
Reproduction first involves getting the testing setup, and then getting the actual project to test against.

### Testing Setup
```
git clone https://github.com/Gregory-Ledray/wasmbrowsertest
cd wasmbrowsertest
git checkout wasme2e
```
Then inspect the script test_build_install.sh for errors for your environment
```
chmod +rwx ./test_build_install.sh
./test_build_install.sh
```

### Project to test against
first go somewhere outside a repo. Then...
```
git clone https://gitlab.com/polyapp-open-source/build-wasm
cd build-wasm
git checkout 1-create-and-run-integration-tests-which-require-a-browser
```
Inspect the test_build_run.sh script for errors in your environment.

This next step will start a convenient web server.
```
./test_build_run.sh
```
This next step is, finally, where the problem is. Note this script makes assumptions about where you installed `wasmbrowsertest`.
```
./test_integration.sh
```

### The Problem
Finally, we can see the problem. If you navigate to localhost:4200 and open dev tools, you see this, which is exactly what I expect to see and is made possible by running the integration testing wasm file when you load the web server started with ./test_build_run.sh (aka there is no chromedp involved - ./test_integration.sh isn't relevant):
```
--- FAIL: TestAlwaysFail (0.00s)
wasm_exec.js:41     main_integration_test.go:20: always fail
wasm_exec.js:41 FAIL
(index):46 go program terminated
wasm_exec.js:487 Uncaught Error: Go program has already exited
    at global.Go._resume (wasm_exec.js:487)
    at wasm_exec.js:500
    at XMLHttpRequest.req.onload ((index):62)
_resume @ wasm_exec.js:487
(anonymous) @ wasm_exec.js:500
req.onload @ (index):62
load (async)
fetchWrapper @ (index):61
syscall/js.valueCall @ wasm_exec.js:327
syscall_js.valueCall @ wasm-0200847e-1263:1
syscall_js.Value.Call @ wasm-0200847e-1249:1
gitlab.com_polyapp_open_source_build_wasm_cmd_frontend.ButtonPressCallback @ wasm-0200847e-2843:1
gitlab.com_polyapp_open_source_build_wasm_cmd_frontend.TestButtonPressCallback @ wasm-0200847e-2848:1
testing.tRunner @ wasm-0200847e-2253:1
wasm_pc_f_loop @ wasm-0200847e-964:1
wasm_export_run @ wasm-0200847e-962:1
run @ wasm_exec.js:478
(anonymous) @ (index):45
Promise.then (async)
(anonymous) @ (index):41
(anonymous) @ (index):53
```

Now consider the output from running the test_integration script in build-wasm, which is running chromedp in headless mode:
```
greg@LAPTOP-AH6IAHBO:/mnt/c/Users/gledr/Polyapp_Apps/gocode/src/gitlab.com/polyapp-open-source/build-wasm$ ./test_integration.sh
integration testing WASM code GOOS=js; GOARCH=wasm
[wasmbrowsertest]: 2020/02/05 12:22:27 main.go:203: fatal error while trying to run tests: {"type":"error","args":[{"type":"object
","subtype":"error","className":"TypeError","description":"TypeError: Could not download wasm module","objectId":"{\"injectedScrip
tId\":2,\"id\":1}","preview":{"type":"object","subtype":"error","description":"TypeError: Could not download wasm module","overflo
w":false,"properties":[{"name":"stack","type":"string","value":"TypeError: Could not download wasm module"},{"name":"message","typ
e":"string","value":"Could not download wasm module"}]}}],"executionContextId":2,"timestamp":1580926947426,"stackTrace":{"callFram
es":[{"functionName":"","scriptId":"4","url":"http://localhost:62603/","lineNumber":51,"columnNumber":20}]}}

```
If you don't immediately see this error, re-run ./test_integration.sh - sometimes it silently fails the first time (no idea why yet).

We see the error is occurring because **the wasm file did not download**. I have no idea why my simple web server is capable of serving the WASM file to a real Chrome instance running on Windows yet this similar clone running inside wasmbrowsertest isn't able to serve to Chromium in headless mode. I'm especially baffled because I basically copied the server from the existing wasmbrowsertest web server in handler.go. For your amusement, you can mess with `func (ws *simpleWebServer) ServeHTTP` in wasmbrowsertest/main.go (which is the web server serving into chromedp) by, say, eliminating the special case for wasm code. Rebuild with the scripts as before and notice you get the same result.

### Summary and Speculation
For some reason my tests work in real Chrome but not when run in chromedp.

One of the following *could* be true:
1. Chromium in WSL seems to be handling a served WASM file differently than Chrome in Windows. This seems unlikely since it works with the regular wasmbrowsertest (aka the tests run with the regular wasmbrowsertest from github.com/agnivade/wasmbrowsertest).
2. chromedp is not setting up Chromium correctly in WSL for it to receive said WASM file. This doesn't really make sense to me because chromedp has loaded the file correctly before, as noted above.
3. There's something wrong with my test setup which leads the WASM file to not be downloaded by my wasmbrowsertest branch, but yet another web server I have works, and wasmbrowsertest 's regular setup also works. After staring at index.html for a while I don't see any important differences when we're calling WebAssembly.instantiateStreaming

If you'd like to stare at index.html (the only relevant HTML), you can do so in `build-wasm/frontend_output/index.html`
Excerpt which is throwing the TypeError:
```
(async() => {
        const go = new Go();
        // pull WASM exit code into app logic
        go.exit = goExit;

        let mod, inst;
        await WebAssembly.instantiateStreaming(fetch("frontend.wasm"), go.importObject).then(async (result) => {
            mod = result.module;
            inst = result.instance;
            document.getElementById("backendButton").disabled = false;
            await go.run(inst);
            console.log('go program terminated');
            // I expect go.run to run forever
            // if we reach here, the app has crashed or the test suite has finished
            document.getElementById("wasm_done").hidden = false;
        }).catch((err) => {
            document.getElementById("wasm_done").hidden = false;
            console.error(err);
        });
    })();
```
Excerpts (incomplete) of the chromedp setup - not shown is the http.FileServer:
```
opts := chromedp.DefaultExecAllocatorOptions[:]
// WSL needs the GPU disabled. See issue #10
if runtime.GOOS == "linux" && isWSL() {
	opts = append(opts,
		chromedp.DisableGPU,
	)
}
done := make(chan struct{})
go func() {
	err = httpServer.Serve(l)
	if err != http.ErrServerClosed {
		logger.Println(err)
	}
	done <- struct{}{}
}()

chromedp.ListenTarget(ctx, func(ev interface{}) {
	handleEvent(ctx, ev, httpServer, done, logger)
})
tasks := []chromedp.Action{
	chromedp.Navigate(`http://localhost:` + port),
	chromedp.WaitEnabled(`#wasm_done`),
	chromedp.Evaluate(`exitCode;`, &exitCode),
}
err = chromedp.Run(ctx, tasks...)
```

**Comment by Gregory-Ledray** [comment]
Alright. I'm now trying to strip down the code as much as possible to narrow down the problem. My browsertest on a different branch has chromedp version;
github.com/chromedp/chromedp v0.5.3

And still has the issue I describe - I'm able to serve the WASM file in Chrome, but not running headless in WSL.

I also realized there are, actually, diffs in the index.html file. The one that's working both in the browser and in chromedp includes go.argv and go.env passed in to instantiateStreaming (adding them didn't solve the problem - frankly it shouldn't be related); I'm calling go.run(inst) in a callback, although the code _appears_ to fail before that (and switching this didn't solve the problem); and I have another script in index.html, but that shouldn't be relevant.

**Comment by Gregory-Ledray** [comment]
I started logging debugging.

```
2020/02/05 16:43:28 browser log
2020/02/05 16:43:28 <- %s
2020/02/05 16:43:28 {"method":"Log.entryAdded","params":{"entry":{"source":"network","level":"error","text":"Failed to load resource: net::ERR_CONTENT_LENGTH_MISMATCH","timestamp":1.580942608985688e+12,"url":"http://localhost:56716/frontend.wasm","networkRequestId"
:"13147.5"}},"sessionId":"D5685E4988AEA0465F14823E3BA7D837"}
2020/02/05 16:43:28 browser log
2020/02/05 16:43:28 <- %s
```

It's a content length mismatch when it tries to download.

**Comment by kenshaw** [comment]
Some notes:

1) I would not recommend / try to run Chrome on Windows Subsystem for Linux (WSL). I would suggest getting this to work properly on either Windows or Linux, separately, before trying it with WSL. You need to isolate whether or not WSL is a factor in this.

2) I really appreciate the writeup and effort here; but in order for me (or others working on `chromedp`) to assist with this issue, we need something significantly shorter and easier to walk through. I will try to look at this, but as it is, but it's not likely in the short term that I'll have time to go through this.

3) This "sounds" to me like an issue with Chrome/ium (and/or running Chrome/ium on WSL) and not something specific to `chromedp`.

**Comment by Gregory-Ledray** [comment]
Hi kenshaw, at this point I don't need people to read everything I wrote, so I just added a note on the initial Issue to skip down to my content length mismatch comment.

Instead of trying to replicate on your own machine, could you just answer this question please? . 
Is there some way to see what the content length of `frontend.wasm` is in chromedp? Even with all of the logging functions on I'm not seeing what chromium is _receiving_, and only able to see what I'm setting in my web server.

**Comment by kenshaw** [comment]
You can enable Chrome's [`Network` domain](https://chromedevtools.github.io/devtools-protocol/tot/Network#method-enable). Events then received by `chromedp` will be displayed, assuming you have set up a [logger](https://godoc.org/github.com/chromedp/chromedp#WithLogf).

You can also try using [`chromedp-proxy`](https://github.com/chromedp/chromedp-proxy), which can work as an intermediate logger (for use with remote hosts, etc).

**Comment by kenshaw** [comment]
Sorry -- if you want to actually see what your web server is sending, then I'd suggest using some kind of intermediate HTTP proxy, which is definitely out of scope of the `chromedp` project.

**Comment by Gregory-Ledray** [comment]
I think the problem has something to do with this:
https://stackoverflow.com/questions/41275918/how-to-get-neterr-content-length-mismatch-in-chrome-using-express
> It looks like **most clients** [emphasis added] don't care if the server sends more data than is stated in Content-Length. However, if you send less data, and make sure that the connection gets closed server-side (the client won't close it because it thinks there's still more data to come), you can trigger net::ERR_CONTENT_LENGTH_MISMATCH

I'd guess either the content-length is getting set wrong or the server is never sending the file.

With Go web servers the content-length is set if content-encoding is not set; if I set both content-encoding to "identity" and content-length to a number, perhaps I'll get a different result.

Either way, I don't think this is chromedp / something chromedp will need to do dev to solve.

Thanks for your suggestions kenshaw! I'll look into the Network domain.

**Comment by Gregory-Ledray** [comment]
Tl;dr if you see an error message like this, consider these things:

1. Open Firefox and real Chrome to your web server & port and see if they say the same things as headless Chrome. If so, you can debug with regular Chrome / Firefox instead of headless.
2. If Chrome and Firefox are saying something different, it could just be a non-standardized error message but the same underlying problem.
3. The Chrome and Firefox's C++ code is often pretty understandable even if you don't remember all the C++ syntax.

Longer explanation:
After doing (1), I found that even though headless Chrome has an ERR_CONTENT_LENGTH_MISMATCH, it's not necessarily an issue with the content length header. I know this because when I tried opening the website in Firefox, I got this instead:
> CompileError: "wasm validation error: at offset 4116005: failed to start custom section"

Which lead me [here ](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmValidate.cpp)to see this code:

```
if (bytesRemain() < (*range)->size) {
  goto fail;
}

CustomSectionEnv sec;
if (!readVarU32(&sec.nameLength) || sec.nameLength > bytesRemain()) {
  goto fail;
}

sec.nameOffset = currentOffset();
sec.payloadOffset = sec.nameOffset + sec.nameLength;

uint32_t payloadEnd = (*range)->start + (*range)->size;
if (sec.payloadOffset > payloadEnd) {
  goto fail;
}
```
```
fail:
  return fail("failed to start custom section");
```
All of this *somewhat* clarifies the error. My content-length is 4,199,262 and the error references 4,116,005 so presumably something claimed it was size X but that was too large for the space left in the .wasm file which had yet to be parsed; thus it threw an error. So this whole thing was some sort of a WASM compilation issue.
